src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -210,44 +240,59 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
                         // the bounds check for the next lookup of span[length]
                         if ((uint)length < (uint)span.Length)
                         {
-                            // Early memory read to hide latency
-                            var expectedCR = span[length];
-                            // Correctly has a CR, move to next
-                            length++;
+                            var headerSpan = span[..length]; 
 
-                            if (expectedCR != ByteCR)
+                            if (length < 4)
                             {
-                                // Sequence needs to be CRLF not LF first.
-                                RejectRequestHeader(span[..length]);
+                                // Less than min possible headerSpan of 3 bytes a:b
+                                RejectRequestHeader(headerSpan);
```

Comment:
Here's the additive way to fix this would be to add the following after the `if (ch1 == ByteCR) { ... }` block above:

```csharp
if (ch1 == ByteCR) {
  // ...
}
else if (_enableLineFeedTerminator && ch1 == ByteLF)
{
    // If we got 2 bytes from the span directly so skip ahead 1 so that
    // the reader's state matches what we expect
    if (readAhead == 0)
    {
        reader.Advance(1);
    }
    // If we got 2 bytes from TryRead, rewind 1 so that
    // the reader's state matches what we expect
    else if (readAhead == 2)
    {
        reader.Rewind(1);
    }
    // If we got exactly 1 byte from TryRead, we're already exactly where we want to be.

    // Double LF found, so end of headers.
    handler.OnHeadersComplete(endStream: false);
    return true;
}
```

That said, I really despise this approach because it adds so much cyclomatic complexity. It's already way out of control for `ParseHeaders` and `ParseMultiSpanHeader`. I think `ParseRequestLine` is simple enough if we take my suggestion from https://github.com/dotnet/aspnetcore/pull/41101/files#r847901254.

We need to refactor this logic to look for "empty" header lines to trigger `OnHeadersComplete` instead of explicitly looking for carriage returns and line feeds. There should be one step at the beginning of `ParseHeaders` and `ParseMultiSpanHeader` that finds the offset for the line terminator whether that's `\r\n` or just `\n` and the rest of the common logic should not take `_enableLineFeedTerminator` into account at all.

The common logic should just parse the header up to the line terminator offset wherever it is, and an empty header should mean the headers have ended. We'd probably still optimize for the case of the new header line starting with a `\r` before we go ahead and call `span.IndexOfAny(ByteCR, ByteLF);`, but that still be part of the beginning step to find the offset for the line terminator that would start the parsing of each header. It would not be combined with the `// Check for final CRLF.` like it is today.

Reply:
I explicitly added tests to fail on `\n\n`, and only accept `\n\r\n` as headers terminator.
Do you really think it should pass?

Reply:
> I explicitly added tests to fail on `\n\n`, and only accept `\n\r\n` as headers terminator. Do you really think it should pass?

I really do. I'm not sure any client actually uses `\n\r\n` to terminate headers. A `\n\n` terminator at least makes some amount of sense even though it's also technically non-compliant. Without accepting `\n\n`, I wonder if this change will fix anything. That's why I asked:

> Can we get a client that uses `\n` instead of `\r\n` for line terminators and manually test it to ensure that Kestrel actually works with the requests these client send? I assume any such client would use `\n\n` to terminate the header sequence. It would be surprising for it to use carriage returns just for that part of the request.

I don't think we should merge this PR until we test with a real client that we know to be broken before this change.